### PR TITLE
feat(apollo explorer): allow callbacks using apiholder

### PR DESCRIPTION
**Pull request by karlhaworth**
## Hey, I just made a Pull Request!

feat(apollo explorer): allow callbacks using apiholder

This will allow callbacks to be used so that tokens can be utilized with Apollo Studio through async apirefs.

Continuation from https://github.com/backstage/backstage/pull/22775. Strategy change recommended here -> https://github.com/backstage/backstage/pull/22775#issuecomment-1931785640.

#### :heavy_check_mark: Checklist

- [X] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [X] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [X] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by karlhaworth** [comment]
@vinzscam or @freben you both made comments on https://github.com/backstage/backstage/pull/22775. This is the follow up from that.

File: plugins/apollo-explorer/src/components/ApolloExplorerPage/ApolloExplorerPage.tsx
Comment (vinzscam) [inline plugins/apollo-explorer/src/components/ApolloExplorerPage/ApolloExplorerPage.tsx:66]:
```suggestion
  }, [endpoints]);
```

**Review by vinzscam — CHANGES_REQUESTED** [review]
nice! added just a tiny suggestion in case the `endpoints` prop changes

**Comment by karlhaworth** [comment]
> nice! added just a tiny suggestion in case the `endpoints` prop changes

@vinzscam Committed the suggestion. Thanks!

**Review by vinzscam — APPROVED** [review]

**Comment by karlhaworth** [comment]
I still want this! @Rugvip

**Comment by karlhaworth** [comment]
Keep open. Waiting for approval.

File: plugins/apollo-explorer/README.md
Comment (Rugvip) [inline plugins/apollo-explorer/README.md:86]:
```suggestion
  const sso = options.apiHolder.get(ssoAuthApiRef)
```

I expect this to work as type should be inferred from the API ref and the type should be correct

File: plugins/apollo-explorer/README.md
Comment (Rugvip) [inline plugins/apollo-explorer/README.md:92]: This token will eventually expire if you stay on the page for too long. We'll need something that's called on every request rather than only when loading the endpoint list.

What I think we'll want to do instead is to forward the `handleRequest` option, where we could also forward the `apiHolder` in the second options argument of that callback. I think there's also a potential security benefit with that since the `headers` option strikes me as something that might be a bit more visible and prone to leakage than using `handleRequest`.

**Review by Rugvip — CHANGES_REQUESTED** [review]
Thank you! 👍 

I think we need to switch up the approach a bit due to expiring tokens, but we do get to remove some of the async handling in the process

**Comment by karlhaworth** [comment]
@backstage/maintainers can you re-run the failing test? Appears to be a test timeout that isn't inside my control or changes.

image

File: plugins/apollo-explorer/README.md
Comment (karlhaworth) [reply plugins/apollo-explorer/README.md]: I'm not sure I followed completely, however [84c4560](https://github.com/backstage/backstage/pull/23405/commits/84c4560c4fa9862a9560bbc7c9431875086b8d84) might do it?

**Comment by freben** [comment]
Done

**Comment by karlhaworth** [comment]
@freben the test is failing again. Anything else I should do? I don't want this to affect the ability to review this PR. Thanks!

**Comment by vinzscam** [comment]
> @freben the test is failing again. Anything else I should do? I don't want this to affect the ability to review this PR. Thanks!

all good let's ship this!

File: plugins/apollo-explorer/README.md
Comment (Rugvip) [inline plugins/apollo-explorer/README.md:81]: Let's use a completely made up thing or a real example, this might trick people into thinking it exists 😅

File: plugins/apollo-explorer/api-report.md
Comment (Rugvip) [inline plugins/apollo-explorer/api-report.md:33]:
```suggestion
}) => Promise<{ token: string }>;
```

General thing for future proofing and helps with readability

File: plugins/apollo-explorer/src/components/ApolloExplorerPage/ApolloExplorerPage.tsx
Comment (Rugvip) [inline plugins/apollo-explorer/src/components/ApolloExplorerPage/ApolloExplorerPage.tsx:57]: This needs to be configured as part of the endpoint right? Seems pretty reasonable that you might need different auth flows for different endpoints. Or perhaps a middle ground to keep things simple is that we forward the `endpointUrl` to the auth callback.

File: plugins/apollo-explorer/src/components/ApolloExplorerBrowser/ApolloExplorerBrowser.tsx
Comment (Rugvip) [inline plugins/apollo-explorer/src/components/ApolloExplorerBrowser/ApolloExplorerBrowser.tsx:73]: Hmm, we're adding this option but hardcoding it to `false` below. Could just drop this bit right?

File: plugins/apollo-explorer/src/components/ApolloExplorerBrowser/ApolloExplorerBrowser.tsx
Comment (Rugvip) [inline plugins/apollo-explorer/src/components/ApolloExplorerBrowser/ApolloExplorerBrowser.tsx:70]:
```suggestion
  authCallback: Props['authCallback'];
```

File: plugins/apollo-explorer/src/components/ApolloExplorerBrowser/ApolloExplorerBrowser.tsx
Comment (Rugvip) [inline plugins/apollo-explorer/src/components/ApolloExplorerBrowser/ApolloExplorerBrowser.tsx:62]:
```suggestion
```

Looks like this is unused

File: plugins/apollo-explorer/src/components/ApolloExplorerBrowser/ApolloExplorerBrowser.tsx
Comment (Rugvip) [inline plugins/apollo-explorer/src/components/ApolloExplorerBrowser/ApolloExplorerBrowser.tsx:89]: This'll need an update for the return type tweak, but you can also skip passing `{}` here, mainly just confusing to the reader 😁

**Review by Rugvip — CHANGES_REQUESTED** [review]
Yep that's what I had in mind! 👍 

Just a bunch of nits 😅

File: plugins/apollo-explorer/src/components/ApolloExplorerBrowser/ApolloExplorerBrowser.tsx
Comment (karlhaworth) [reply plugins/apollo-explorer/src/components/ApolloExplorerBrowser/ApolloExplorerBrowser.tsx]: [a1e39ff](https://github.com/backstage/backstage/pull/23405/commits/a1e39ffd2a711b6eb3129c497b2e2b11f059e429) [6ecacd3](https://github.com/backstage/backstage/pull/23405/commits/6ecacd3ffec0a25663803337363ad59d0d189906)

File: plugins/apollo-explorer/src/components/ApolloExplorerPage/ApolloExplorerPage.tsx
Comment (karlhaworth) [reply plugins/apollo-explorer/src/components/ApolloExplorerPage/ApolloExplorerPage.tsx]: [3f422fb](https://github.com/backstage/backstage/pull/23405/commits/3f422fbd66bc0262d9df60c7f8c5a6ca5fa182c0) should cover this

File: plugins/apollo-explorer/README.md
Comment (karlhaworth) [reply plugins/apollo-explorer/README.md]: [8ea42d7](https://github.com/backstage/backstage/pull/23405/commits/8ea42d759e040ff3f22f9fefdf20d01171994780) `companyxyz` work?

**Comment by karlhaworth** [comment]
@Rugvip think I've taken care of all your comments! 😄

**Review by Rugvip — APPROVED** [review]
Alright, let's :shipit:!

**Pull request merged**
